R/bootstrap.R

```diff
@@ -145,20 +141,19 @@ renv_bootstrap_repos_lockfile <- function() {
 
 renv_bootstrap_download <- function(version) {
 
-  # begin collecting different methods for finding renv
+  sha <- attr(version, "sha", exact = TRUE)
   methods <- c(
-    renv_bootstrap_download_tarball,
-    # dev versions can only come from GitHub
-    if (renv_bootstrap_version_is_dev(version))
-      renv_bootstrap_download_github
+    function() renv_bootstrap_download_tarball(version),
+    if (!is.null(sha))
+      function() renv_bootstrap_download_github(sha)
     else c(
-      renv_bootstrap_download_cran_latest,
-      renv_bootstrap_download_cran_archive
+      function() renv_bootstrap_download_cran_latest(version),
+      function() renv_bootstrap_download_cran_archive(version)
     )
   )
```

Comment:
I wonder if it's better to move the `if` up a level:

```R
methods <- if (is.null(sha)) {
  c(
   function() renv_bootstrap_download_tarball(version),
   function() renv_bootstrap_download_github(sha)
  )
} else {
  c(
    function() renv_bootstrap_download_tarball(version),
    function() renv_bootstrap_download_cran_latest(version),
    function() renv_bootstrap_download_cran_archive(version)
  )
}
```

Or possibly even check `renv_envvar_exists("RENV_BOOTSTRAP_TARBALL")` first, because I feel like if that's set and the file doesn't exist, you probably don't want to fall back to the default.